pkg/agent/multicast/vif_allocator.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package multicast
+
+import (
+	"container/list"
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/vishvananda/netlink"
+	"k8s.io/klog/v2"
+)
+
+// VIF stands for virtual interface for multicast in kernel multicast implementation
+// and is used for adding/deleting static multicast entries, please check
+// https://github.com/torvalds/linux/blob/master/include/uapi/linux/mroute.h#L80
+// VIFAllocator is used for saving vif <-> interface mapping.
+// It also allocates and deallocates VIFs because VIFs are limited resource(at most 32)
+// for kernel prior to 5.10
+type VIFAllocator struct {
+	mutex         sync.RWMutex
+	vifMap        map[string]uint16
+	vifReverseMap map[uint16]string
+	nextID        uint16
+	maxID         uint16
+	sockFD        int
+	availableIDs  *list.List
+}
+
+func newVIFAllocator(sockFD int) *VIFAllocator {
+	return &VIFAllocator{
+		availableIDs:  list.New(),
+		vifMap:        make(map[string]uint16),
+		vifReverseMap: make(map[uint16]string),
+		nextID:        0,
+		maxID:         255,
+		sockFD:        sockFD,
+	}
+}
+
+func (v *VIFAllocator) getVif(name string) (uint16, error) {
+	v.mutex.RLock()
+	defer v.mutex.RUnlock()
+	vif, ok := v.vifMap[name]
+	if !ok {
+		return 0, fmt.Errorf("VIF for interface %s not found", name)
+	}
+	return vif, nil
+}
+
+func (v *VIFAllocator) allocateVif(name string) (uint16, error) {
+	l, err := netlink.LinkList()
+	if err != nil {
+		return 0, err
+	} else {
```

Comment:
no need to add block "else", you can write the code directly because the function is already returned if err!=nil